bgpd/bgp_vty.c

```diff
@@ -6159,21 +6159,46 @@ static int set_ecom_list(struct vty *vty, int argc, struct cmd_token **argv,
 	return CMD_SUCCESS;
 }
 
-static int vpn_policy_getafi(struct vty *vty, int *doafi)
+/*
+ * v2vimport is true if we are handling a `import vrf ...` command
+ */
+static afi_t vpn_policy_getafi(struct vty *vty, struct bgp *bgp, bool v2vimport)
 {
+	afi_t afi;
+
 	switch (vty->node) {
 	case BGP_IPV4_NODE:
-		doafi[AFI_IP] = 1;
+		afi = AFI_IP;
 		break;
 	case BGP_IPV6_NODE:
-		doafi[AFI_IP6] = 1;
+		afi = AFI_IP6;
 		break;
 	default:
 		vty_out(vty,
 			"%% context error: valid only in address-family <ipv4|ipv6> unicast block\n");
-		return CMD_WARNING_CONFIG_FAILED;
+		return AFI_MAX;
 	}
-	return CMD_SUCCESS;
+
+	if (!v2vimport) {
+		if (CHECK_FLAG(bgp->af_flags[afi][SAFI_UNICAST],
+			       BGP_CONFIG_VRF_TO_VRF_IMPORT)
+		    || CHECK_FLAG(bgp->af_flags[afi][SAFI_UNICAST],
+				  BGP_CONFIG_VRF_TO_VRF_EXPORT)) {
+			vty_out(vty,
+				"%% error: Please unconfigure import vrf commands before using vpn commands\n");
+			return AFI_MAX;
+		}
+	} else {
+		if (CHECK_FLAG(bgp->af_flags[afi][SAFI_UNICAST],
+			       BGP_CONFIG_VRF_TO_MPLSVPN_EXPORT)
+		    || CHECK_FLAG(bgp->af_flags[afi][SAFI_UNICAST],
+				  BGP_CONFIG_MPLSVPN_TO_VRF_IMPORT)) {
+			vty_out(vty,
+				"%% error: Please unconfigure vpn to vrf commands before using import vrf commands\n");
+			return AFI_MAX;
+		}
+	}
+	return afi;
```

Comment:
This new block above looks good as an interlock to prevent accidental interference between the two leaking modes. I wonder if you'd consider making it into a separate function as it is unrelated to the vty node afi lookup.

Reply:
I put it in this function because it was going to be called everywhere that we cared and if we added new functionality in the future it probably would be called to and thus continuing to work.